code/game/objects/structures/popout_cake.dm

```diff
@@ -0,0 +1,185 @@
+			do_popout()
+			if(!strong_surprise)
+				for(var/i=1 to (amount_of_slices))
+					var/obj/item/reagent_containers/food/snacks/slice = new slice_path (loc)
+					slice.initialize_slice(slice, 0)
+			qdel(src)
+			return FALSE
+	if(istype(W, /obj/item/grenade/flashbang))
+		if(strong_surprise)
+			to_chat(user, "<span class='notice'>There's no space for [src] inside!</span>")
+		else
+			user.visible_message("<span class = 'notice'>[user] begins inserting [W] into [src]!</span>", "<span class = 'notice'>You begin inserting [W] into [src]!</span>")
+			if(do_after(user, 30, FALSE, src))
+				strong_surprise = TRUE
+				user.visible_message("<span class = 'notice'>After some fiddling, [user] inserts [W] into [src]!</span>", "<span class = 'notice'>You attach [W] to the hidden mechanism inside!</span>")
+				qdel(W)
+				return FALSE
+	else
+		..()
+
+/obj/structure/popout_cake/proc/do_popout()
+	if(isnull(occupant))
+		return
+	visible_message("<span class = 'notice'>Loud shuffling can be heard from inside [src]!</span>")
+	if(!used_string)
+		used_string = TRUE
+		playsound(src, 'sound/items/party_horn.ogg', 50, 1)
+		var/datum/effect_system/spark_spread/s = new()
+		s.set_up(6, FALSE, src)
+		s.start()
+	if(strong_surprise) //This is the extra OOMPH, a mini flashbang plus a small explosion that scatters all the pie slices around!
+		var/flashing_turf = get_turf(src)
+		for(var/mob/living/M in viewers(5, flashing_turf))
+			if(M != occupant) //So that the guy hiding inside doesn't get flashed
+				flash_and_bang(get_turf(M), M)
+				for(var/i=1 to (amount_of_slices))
+					var/obj/item/reagent_containers/food/snacks/slice = new slice_path (loc)
+					slice.initialize_slice(slice, 0)
+					slice.throw_at(get_edge_target_turf(src,pick(GLOB.alldirs)),rand(1,3),5)
+	string.Remove(occupant)
+	occupant.forceMove(get_turf(src))
+	occupant = null
+	if(strong_surprise)
+		qdel(src)
+
+/obj/structure/popout_cake/proc/flash_and_bang(turf/T, mob/living/M)
+	if(M.stat == DEAD)	//They're dead!
+		return
+	var/distance = max(0,get_dist(get_turf(src),T))
+	//When distance is 0, will be 1
+	//When distance is 5, will be 0
+	//Can be less than 0 due to hearers being a circular radius.
+	var/distance_proportion = max(1 - (distance / 5), 0)
+	M.show_message("<span class='warning'>BANG!</span>", MSG_AUDIBLE)
+	if(M.flash_act(intensity = 1, affect_silicon = 1))
+		if(distance_proportion)
+			M.Paralyze(20 * distance_proportion)
+			M.Knockdown(200 * distance_proportion)
+	else
+		M.flash_act(intensity = 2)
+	if(!distance || loc == M || loc == M.loc)	//Stop allahu akbarring rooms with this.
+		M.Paralyze(20)
+		M.Knockdown(200)
+		M.soundbang_act(1, 200, 10, 15)
+	else
+		if(distance <= 1)
+			M.Paralyze(5)
+			M.Knockdown(30)
+		distance_proportion = max(1 - (distance / 5), 0)
+		if(distance_proportion)
+			M.soundbang_act(1, 200 * distance_proportion, rand(0, 5))
+
+/obj/structure/popout_cake/Destroy()
+	if(occupant)
+		do_popout()
```

Comment:
You need to remove the reference to the cake from the string action or else it will harddel
`string.cake = null`